### PR TITLE
Improve reduce_dims for speed up

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

File: cupy/core/elementwise.pxi [hunk 1]
```diff
@@ -211,58 +211,64 @@ cpdef tuple _reduce_dims(list args, tuple params, tuple shape):
...
+        cnt = 1
         axis = ndim - 1
+        total_size = internal.prod(shape)
 
-    if cnt == ndim:
-        return args, shape
     if cnt == 1:
-        newshape.assign(<Py_ssize_t>1, <Py_ssize_t>vecshape[axis])
+        newshape.assign(<Py_ssize_t>1, total_size)
         ret = []
```
Comment (toslunar) [inline cupy/core/elementwise.pxi]: `ret` is no longer used.

File: cupy/core/elementwise.pxi [hunk 2]
```diff
@@ -211,58 +211,64 @@ cpdef tuple _reduce_dims(list args, tuple params, tuple shape):
...
-                a = arr
-            ret.append(a)
-        return ret, tuple(newshape)
+                args[i] = arr
+        return total_size,
 
     for i in range(ndim):
         if vecshape[i] != 1:
             newshape.push_back(vecshape[i])
     ret = []
```
Comment (toslunar) [inline cupy/core/elementwise.pxi]: ditto

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
In the above example, `args[0]` is not c-contiguous, `args[1]` is c-contiguous, and the broadcasted shape is `(3, 1, 2)`.  We should have `axes == [0, 2]` (not `[1, 2]`).  While strides of c-contiguous arg don't have to be checked, `args[1].stride[1]` cannot be used since `shape[1] == 1`.

**Comment by okuta** [comment]
Thanks, I added test.

File: cupy/core/elementwise.pxi [hunk 3]
```diff
@@ -172,78 +172,71 @@ cpdef str _get_kernel_params(tuple params, tuple args_info):
 
 
 cpdef tuple _reduce_dims(list args, tuple params, tuple shape):
-    cdef Py_ssize_t i, j, n, ndim, cnt, axis, s
+    """ Remove contiguous stride to optimize CUDA kernel."""
+    cdef int i, j, ax, last_ax, ndim
```
Comment (toslunar) [inline cupy/core/elementwise.pxi]: - Why did you change `Py_ssize_t` to `int`?
- `j` is no longer used.

File: cupy/core/elementwise.pxi [hunk 4]
```diff
@@ -172,78 +172,71 @@ cpdef str _get_kernel_params(tuple params, tuple args_info):
...
+    cdef Py_ssize_t total_size
     cdef vector.vector[Py_ssize_t] vecshape, newshape, newstrides
-    cdef vector.vector[bint] is_array_flags
+    cdef vector.vector[int] array_indexes, axes
```
Comment (toslunar) [inline cupy/core/elementwise.pxi]: ditto

**Comment by okuta** [comment]
Because `int` has enough representation.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.
